ksql-streams/src/test/java/io/confluent/ksql/execution/streams/StreamSelectKeyBuilderTest.java

```diff
@@ -0,0 +1,235 @@
+        false
+    );
+  }
+
+  @Test
+  @SuppressWarnings("unchecked")
+  public void shouldRekeyCorrectly() {
+    // When:
+    final KStream result = StreamSelectKeyBuilder.build(kstream, selectKey);
+
+    // Then:
+    verify(kstream).filter(any());
+    verify(filteredKStream).selectKey(any());
+    verify(rekeyedKstream).mapValues(any(ValueMapperWithKey.class));
+    verifyNoMoreInteractions(kstream, filteredKStream, rekeyedKstream, updatedKeyKStream);
+    assertThat(result, is(updatedKeyKStream));
+  }
+
+  @Test
+  @SuppressWarnings("unchecked")
+  public void shouldFilterOutNullValues() {
+    // When:
+    StreamSelectKeyBuilder.build(kstream, selectKey);
+
+    // Then:
+    verify(kstream).filter(predicateCaptor.capture());
+    final Predicate<Object, GenericRow> predicate = predicateCaptor.getValue();
+    assertThat(predicate.test(asStructKey("dre"), null), is(false));
+  }
+
+  @Test
+  @SuppressWarnings("unchecked")
+  public void shouldFilterOutNullKeyColumns() {
+    // When:
+    StreamSelectKeyBuilder.build(kstream, selectKey);
+
+    // Then:
+    verify(kstream).filter(predicateCaptor.capture());
+    final Predicate<Object, GenericRow> predicate = predicateCaptor.getValue();
+    assertThat(
+        predicate.test(asStructKey("dre"), new GenericRow(0, "dre", 3000, null)),
+        is(false)
+    );
+  }
+
+  @Test
+  @SuppressWarnings("unchecked")
+  public void shouldNotFilterOutNonNullKeyColumns() {
+    // When:
+    StreamSelectKeyBuilder.build(kstream, selectKey);
+
+    // Then:
+    verify(kstream).filter(predicateCaptor.capture());
+    final Predicate<Object, GenericRow> predicate = predicateCaptor.getValue();
+    assertThat(
+        predicate.test(asStructKey("dre"), new GenericRow(0, "dre", 3000, "bob")),
+        is(true)
+    );
+  }
+
+  @Test
+  @SuppressWarnings("unchecked")
+  public void shouldIgnoreNullNonKeyColumns() {
+    // When:
+    StreamSelectKeyBuilder.build(kstream, selectKey);
+
+    // Then:
+    verify(kstream).filter(predicateCaptor.capture());
+    final Predicate<Object, GenericRow> predicate = predicateCaptor.getValue();
+    assertThat(predicate.test(asStructKey("dre"), new GenericRow(0, "dre", null, "bob")), is(true));
+  }
+
+  @Test
+  @SuppressWarnings("unchecked")
+  public void shouldComputeCorrectKey() {
+    // When:
+    StreamSelectKeyBuilder.build(kstream, selectKey);
+
+    // Then:
+    verify(filteredKStream).selectKey(keyValueMapperCaptor.capture());
+    final KeyValueMapper<Object, GenericRow, Struct> keyValueMapper =
+        keyValueMapperCaptor.getValue();
+    assertThat(
+        keyValueMapper.apply(asStructKey("dre"), new GenericRow(0, "dre", 3000, "bob")),
+        is(asStructKey("bob"))
+    );
+  }
+
+  @Test
+  @SuppressWarnings("unchecked")
+  public void shouldUpdateRowkeyIfUpdateRowkeyTrue() {
+    // When:
+    StreamSelectKeyBuilder.build(kstream, selectKey);
+
+    // Then:
+    verify(rekeyedKstream).mapValues(mapperCaptor.capture());
```

Comment:
consider moving into method to make tests more readable:

```java
    private ValueMapperWithKey<Object, GenericRow, GenericRow> getMapper() { 
       verify(rekeyedKstream).mapValues(mapperCaptor.capture());
       final ValueMapperWithKey<Object, GenericRow, GenericRow> mapper = mapperCaptor.getValue();
    }
```

Same for others.

Reply:
consider moving into method to make tests more readable:

```java
    private ValueMapperWithKey<Object, GenericRow, GenericRow> getMapper() { 
       verify(rekeyedKstream).mapValues(mapperCaptor.capture());
       final ValueMapperWithKey<Object, GenericRow, GenericRow> mapper = mapperCaptor.getValue();
    }
```

Same for others.